docs/compatibility.md

```diff
@@ -0,0 +1,183 @@
+# StableHLO Compatibility
+
+## Compatibility Guarantees
+
+(G1) libStablehlo provides 6 months of backward compatibility, i.e. StableHLO programs serialized by an old version of libStablehlo have the same semantics when deserialized by new versions of libStablehlo within the compatibility window.
+
+(G2) libStablehlo provides 3 weeks of forward compatibility, i.e. StableHLO programs serialized by a new version of libStablehlo have the same semantics when deserialized by old versions of libStablehlo within the compatibility window, unless they are using new features introduced since the old version.
```

Comment:
Ah - I didn't realize there are no stability guarantees from the bytecode infrastructure. I suppose the protocols in this proposal are generic enough to apply to "some stable MLIR serialization mechanism", but it would be very nice to use the built-in infrastructure.

Is there any existing discussion surrounding guarantees on the MLIR bytecode infrastructure? Any reasons why it would be undesirable, aside from code maintenance?

Reply:
We will likely stabilize the bytecode serialization "at some point", but it's not a given that it would follow exactly your plan (6 month back, 3 weeks forward) for example.
Even more so in terms of the dialects provided by upstream (like the Arith dialect).

Reply:
Got it, thanks for the info there. Let me keep thinking on that one, will likely reach out to you soon.

As for changes in upstream dialects I agree that more thought is needed. Linking a comment I left in reply to Sean, the current scheme is tolerant of certain types of changes in upstream dialects, migration changes where there is a semantically identical alternative (i.e. Op moves / renames) as well as additions (i.e. can provide forward compatibility warnings), more thought is required for handling deprecations and larger semantic changes:
https://github.com/openxla/stablehlo/pull/115#discussion_r970098874

Reply:
RFC v2 proposes to fork all required operations/attrs/types in order to control the spec on the features we wish to offer compatibility guarantees for.

We still need guarantees from the serialization layer. Hoping to discuss this at the MLIR Summit tomorrow. Worst case, we would need to fork the upstream serialization/deserialization, or create our own serialization machinery if that is not possible. I am planning to work on a Serialization RFC in the coming weeks to solidify a proposal for StableHLO serialization, these points will be addressed in more detail in that RFC.